### DISCUSSION TITLE
关于mortal v3的强度和副露倾向

### DISCUSSION THREAD

**Discussion by Pragmatism0220** [Equim-chan/Mortal#38] [Q&A]
主要有一些疑惑：
1. v3版本的mortal与之前版本相比，强度上是确实提升了、还是整体相当呢？
2. 像饼、呆呆这些麻将选手，他们整体的风格是“副露防守流”，特点是“较高的副露率（往往超过40%）、较低的放铳率（11~13%）、相对较低的打点→相对更高和牌率”。mortal是否有可能训练成这种“副露防守流”的风格呢？
3. 关于训练的疑惑：我目前的训练方式是离线+在线。设置了较大的batch size，在在线训练时则降低了lr，但是往往吃个饭或者睡个午觉之后模型就开始莫名其妙退步，pt变负、顺位变后，有没有什么训练的技巧？

如有打扰非常抱歉！

**Comment by Equim-chan** [comment]
1. 我这边所有有关具体模型间强度的论点均基于 1v3/3v1 test play 后的统计数据，见[文档](https://mortal.ekyu.moe/perf/strength.html)。不过准确地说这些只是 training session 间的强度差异，单纯问版本间的强度差异，我也不知道。
2. 我没有这么尝试过，你可以研究研究。
3. 较大是多大？降低是多低？offline 转 online 还改了哪些配置？描述得比较含糊。另外我想提一下，online 的确很难训练，很折磨，即便能练出来比 offline 更强的，在人类玩家看来很多细节反而像是退步了的。我觉得你可以多观察观察这些，或者说如果 online 在你那确实行不通，可以试着从别的方向去改进（欢迎 PR）。

**Reply by Pragmatism0220** [reply]
batch我尝试了1024\~2048，lr尝试降低到1e6\~5e7……效果不大行

**Reply by Equim-chan** [reply]
我不建议直接改 batch size，在开始训练之后应该要一直保持一致。batch size 的调整可以通过改 opt_step_every 来变相地完成。另外 online 要求 freeze bn。

**Reply by Pragmatism0220** [reply]
好的，我试试。opt_step_every大概提高到多少比较好呢？此外，还遇到的一个现象是，离线训练出的模型，每一打的权重非常极端，基本上都是某个牌倾向99%，其余的都是0点几，偶尔也不过是纠结某两个牌。而我注意到，您的在线网页上的模型，每一打的权重分布在多张牌上，并突出了一个最优解，这是什么原因呢？

**Reply by Equim-chan** [reply]
权重极端这个是 offline (CQL) 的特色，我网站上的模型权重分布均匀是因为那是 offline 转 online 的模型。

**Reply by Pragmatism0220** [reply]
你好，我这边遇到一个新的疑惑。当offline转online的时候，我需要给当前模型一个baseline用作在线训练。我之前offline训练的时候baseline是一个强度尚可的v2模型，但是这次offline训练得到的v3模型从1v3的表现来看，貌似已经**超过**了作为baseline的v2模型。那在这种情况下，如果我想对这个v3模型作online在线训练，我该选取谁作为baseline呢？
我有考虑将v3自己拷贝一份作为baseline，但这样训练得到的结果应该会很不合理，估计会导致模型退化；但是如果以v2模型再作为baseline，那此时v2模型已经统计学意义上弱于v3了，真的还有作为baseline的价值吗？所以想请问一下，在遇到这种**无法选择一个更优的baseline**的情况下，您是怎么做的呢？~~感觉有点类似蛋生鸡还是鸡生蛋的意味了（~~

**Reply by Equim-chan** [reply]
你实际试一下就会发现这些细节都不重要。假如你一开始就没训练这个 v3，想从之前最强的 v2 开始做 online，不还是得取它自己做 baseline 吗。

**Reply by Pragmatism0220** [reply]
好像有道理……谢谢解答，我试试看！

**Reply by Pragmatism0220** [reply]
您好我再次打扰一下，仍然是最开始提问里提到的副露率问题，我的模型副露率最高好像也就31%左右，根据训练方式的不同甚至会更低。我上次提问后又想了一下，既然无模型的AI会认为副露是一种低期望的行为（也许是认为降低了打点的同时还降低了防御力），所以Mortal在不断迭代训练后总是期望门清。我搜索了一下Suphx刚升10段时的副露率，5k多场31%副露，非常门清的打法。
最开始的时候我考虑直接给副露加上某种reward，但我仔细想想认为这非常的不科学，因为副露并不是多多益善，更不是“按钮亮了就点”。单单以此作为reward可能会带来潜在的更大的负收益。
但是考虑一下人类玩家学习副露的思路，往往会以“提升手牌的速度”这个角度。这里的“速度”不仅仅是向听数前进这么片面，包括“故意退向、食延、破一发、错海底、bluff”等等多种因素，像向听数前进、食延这种属于“提升自己的做牌速度”，而“破一发海底、bluff”，则有点类似“绞杀”的思想，属于“拖慢别人的做牌速度或者降低别人的期望打点”来给自己带来正收益的行为。但是这种行为在Mortal和Suphx的论文中我貌似没有看到应予的reward。
我在想，这种行为没有对应的奖励会不会是造成AI不愿意承担副露风险的原因呢？在grp中的奖励如果我没理解错，是以排名和得分作为计算依据、而不依赖具体牌面的。但这个维度上是否没有考虑“**速度**”这个维度呢（也就是类似玩家的平均听牌巡目、平均和了巡目这种指标）？毕竟，AI学会了要去提高打点、提升顺位，但是再大的牌和不了也没有意义啊。单单靠grp的指导貌似导致AI倾向门清凹牌并弃和了。所以我在想，将这种行为纳入考虑是否合理？还是会视作人为的干预呢？
我所想的一种简化的可能的方式，比如，按照巡目，越早听牌，速度权重的reward就越高。但序盘、中盘和终盘的权重不能视作线性递减。最后计算reward时，将速度也作为reward的一部分。这样AI是否就会去计算、权衡副露和打点↓、防御力↓的利弊了呢？
以上只是一个非常粗略的想法，甚至很可能是错的，但还是很想讨论一下，谢谢！

**Reply by Equim-chan** [reply]
你想的这些我也都想过，有些也试过，比如单独给副露一些奖励，或者基于和牌速度给加成，但是最终没有采用，因为效果没有达到预期。最早我尝试用 Suphx 的方式区分吃碰模型的时候也试过例如 focal loss 的方法，当然也都是不管用的。

从理论上说现在的模式并不是没有考虑和牌速度这些因素，同条件下早和牌的 $\hat Q^\pi$ (Mortal 推理的结果)是应该比晚和牌的高的，因为对局如果拉长的话，横移动、放铳、流局这些结果的概率就会增加。另一方面 MDP 的 $\gamma$ 在理论上也能控制这个事情，虽然现在用的是 1，我以前尝试过别的值。 $\gamma$ 会直接介入目标 returns (来自 GRP)，我们的 reward 很稀疏，只有局末尾有一个 reward，其他位置都是 0，而且没有用 TD 而是用了 MC，所以 returns 直接就是 $r \cdot \gamma^s$ ，s 为剩余步数。具体来说，对于同样的初始状态和同样的终局点数 delta，当这局 reward 为正时(例如大多数的和牌)，局长度越长，returns 越低，而这局 reward 为负时(例如大多数的放铳)，局长度越长，returns 越高。

虽然这说法是有点劝退，不过我做麻将 AI 时经常能感到这种无理——很多在理论上很 fancy、很 robust、也很流行的东西，实际在麻将 AI 上实现之后发现都不管用或者没有达到预期，反而是简单粗暴、看上去问题很多、理论上常常说不通的实现能维持一个不错的表现。我原本是要在文档的 failed attempts 里补充这些的，但是一直没什么动力去写。

**Reply by Equim-chan** [reply]
此外还有另一个问题，不是说这些人为干预的 reward 规则不好，而是他们很难制定，甚至对麻将来说如果这个规则能制定出来的话那么它本身都可以作为一个 rule-based AI 了。这些事情需要 domain knowledge，而这是我不太喜欢的（或者说不太符合我个人的哲学）。我认为 ML 的一个重要目的就是为了降低甚至消去 domain knowledge 的成本和可能存在的 bias。

如果你想试着弄的话，可以参考 OpenAI 的 five 是怎么做的，在 Appendix G。

顺带提一句，我觉得大型 model-free RL 这方面目前做得最出色最详实的研究就是 OpenAI 的 five，简直就是大型 model-free RL 的圣经，我时不时都会翻出来看看。

**Reply by Pragmatism0220** [reply]
感谢解答……看起来这并不是一个很容易解决的问题啊，我的想法还是太简单了（捂脸）……受益良多，非常感谢您的解答！我会去看看openAI的相关研究的！

**Reply by Nitasurin** [reply]
@Equim-chan 我以前取过几个深度强化学习常见的值和自己根据一些麻将数据计算出的值，代入gamma作offline转online train，但在长期看效果均不佳。您提到“ 虽然现在用的是1，我以前尝试过别的值”，是有什么考量嘛？

**Reply by Equim-chan** [reply]
@Niitsune 是实验的结果。以前尝试过别 $\gamma$ 值但效果不佳，反而 $\gamma$ = 1 虽然有点违背 MDP 的理论，但是效果不差。我的猜测是麻将里的 episode length 较难预测， $\gamma \in (0, 1)$ 的话会导致 discounted returns 的 variance 变大，反而不方便模型去估计。

**Reply by teble** [reply]
@Equim-chan 关于文档中mortal与akochan对抗时的配置，有修改pt分布等配置吗，我感觉ako的选择貌似与pt分布有很大关系，负和场选择更加谨慎，目前我测试使用的配置是mjai-reviewer中附带的tactics.json。

**Reply by Equim-chan** [reply]
@teble 有啊，文档里就分了 2 种不同 pt 分布的 akochan。

**Reply by teble** [reply]
> @teble 有啊，文档里就分了 2 种不同 pt 分布的 akochan。

感谢解答，我以为那个只是最后计分pt分布，而不是代表akochan的配置pt。
